pkg/pillar/cmd/watcher/watcher.go

```diff
@@ -156,6 +243,180 @@ func handleMemoryPressureEvents() {
+	var sum, sumSq float64
+	for _, value := range data {
+		sum += value
+		sumSq += value * value
+	}
+	mean = sum / n
+	variance := (sumSq / n) - (mean * mean)
+	stdDev = math.Sqrt(variance)
+	return
+}
+
+// detectGoroutineLeaks detects if there's a potential goroutine leak over time.
+// Returns true if a leak is detected, false otherwise.
+func detectGoroutineLeaks(stats []int) (bool, []float64) {
+
+	if len(stats) < 10 {
+		// Not enough data to determine trend
+		return false, nil
+	}
+
+	// The window size for the moving average
+	windowSize := len(stats) / 10
+
+	// Step 1: Smooth the data
+	smoothedData := movingAverage(stats, windowSize)
+
+	if len(smoothedData) < 2 {
+		// Not enough data to determine trend
+		return false, smoothedData
+	}
+
+	// Step 2: Calculate the rate of change
+	rateOfChange := make([]float64, len(smoothedData)-1)
+	for i := 1; i < len(smoothedData); i++ {
+		rateOfChange[i-1] = smoothedData[i] - smoothedData[i-1]
+	}
+
+	// Step 3: Calculate mean and standard deviation of the rate of change
+	mean, stdDev := calculateMeanStdDev(rateOfChange)
+
+	// Step 4: Determine the dynamic threshold
+	threshold := 0.0 + stdDev
+
+	// Step 5: Check if the latest rate of change exceeds the threshold
+	latestChange := rateOfChange[len(rateOfChange)-1]
+	if mean > threshold && latestChange > threshold {
+		log.Warnf("Potential goroutine leak detected: latest increase of %.2f exceeds dynamic threshold of %.2f.", latestChange, threshold)
+		return true, smoothedData
+	}
+	return false, smoothedData
+}
+
+func handlePotentialGoroutineLeak() {
+	// Dump the stack traces of all goroutines
+	agentlog.DumpAllStacks(log, agentName)
+}
+
+// goroutinesMonitor monitors the number of goroutines and detects potential goroutine leaks.
+func goroutinesMonitor(ctx *watcherContext) {
+	// Get the initial goroutine leak detection parameters to create the stats slice
+	goroutinesThreshold, checkInterval, checkStatsFor, keepStatsFor, cooldownPeriod := ctx.GRLDParams.Get()
+	entriesToKeep := int(keepStatsFor / checkInterval)
+	stats := make([]int, 0, entriesToKeep+1)
+	var lastLeakHandled time.Time
+	for {
+		goroutinesThreshold, checkInterval, checkStatsFor, keepStatsFor, cooldownPeriod = ctx.GRLDParams.Get()
+		// Check if we have to resize the stats slice
+		newEntriesToKeep := int(keepStatsFor / checkInterval)
+		if newEntriesToKeep != entriesToKeep {
+			if newEntriesToKeep > entriesToKeep {
+				// **Increasing Size**
+				if cap(stats) >= newEntriesToKeep+1 {
+					// Capacity is sufficient; update entriesToKeep
+					log.Functionf("Capacity sufficient; update entriesToKeep: %d", newEntriesToKeep)
+					entriesToKeep = newEntriesToKeep
+				} else {
+					// Capacity insufficient; create a new slice
+					log.Functionf("Capacity insufficient; create a new slice: %d", newEntriesToKeep+1)
+					newStats := make([]int, len(stats), newEntriesToKeep+1)
+					copy(newStats, stats)
+					stats = newStats
+					entriesToKeep = newEntriesToKeep
```

Comment:
Why do you need to create a new slice?

Later you use `append` anyways.

Reply:
It's a slice of new capacity. It's either bigger or less than the original.

Reply:
I even have unit tests for both cases!

Reply:
But why do you need a bigger slice if you make it bigger with `append` anyways?

Reply:
Mmm... When the capacity is insufficient, does appending cause an implicit allocation?

Reply:
Yes

Reply:
Ok, I'll simplify the code. Thanks for pointing to it!

Reply:
https://go.dev/play/p/3L14ni03ioZ

Reply:
Yeah, I got it. The only minor concern then is that it can allocate too much

https://go.dev/play/p/czCzIfi1QIs

But who cares about mem in Go, right? =))

Reply:
Done